blockchain/stakeversion.go

```diff
@@ -0,0 +1,359 @@
+// Copyright (c) 2016 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package blockchain
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/decred/dcrd/chaincfg/chainhash"
+)
+
+var (
+	errStakeVersionHeight           = errors.New("not enough height")
+	errVoterVersionMajorityNotFound = errors.New("voter version majority " +
+		"not found")
+	errStakeVersionMajorityNotFound = errors.New("stake version majority " +
+		"not found")
+)
+
+// calcWantHeight calculates the height of the final block of the previous
+// interval.  The adjusted height accounts for the fact the starting validation
+// height does not necessarily start on an interval and thus the intervals
+// might not be zero-based.
+func calcWantHeight(skip, interval, height int64) int64 {
+	intervalOffset := skip % interval
+	adjustedHeight := height - intervalOffset - 1
+	return (adjustedHeight - ((adjustedHeight + 1) % interval)) +
+		intervalOffset
+}
+
+// findStakeVersionPriorNode walks the chain backwards from prevNode until it
+// finds the appropriate interval block to return to the caller.
+//
+// This function MUST be called with the chain state lock held (for writes).
+func (b *BlockChain) findStakeVersionPriorNode(prevNode *blockNode) (*blockNode, error) {
+	// Check to see if the blockchain is high enough to begin accounting
+	// stake versions.
+	nextHeight := prevNode.height + 1
+	if nextHeight < b.chainParams.StakeValidationHeight+
+		b.chainParams.StakeVersionInterval {
+		return nil, errStakeVersionHeight
+	}
+
+	wantHeight := calcWantHeight(b.chainParams.StakeValidationHeight,
+		b.chainParams.StakeVersionInterval, nextHeight)
+
+	// Walk backwards until we find an interval block and make sure we
+	// don't blow through the minimum height.
+	iterNode := prevNode
+	for iterNode.height > wantHeight {
+		var err error
+		iterNode, err = b.getPrevNodeFromNode(iterNode)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return iterNode, nil
+}
+
+// isVoterMajorityVersion determines if minVer requirement is met based on
+// prevNode.  The function always uses the voter versions of the prior window.
+// For example, if StakeVersionInterval = 11 and StakeValidationHeight = 13 the
+// windows start at 13 + (11 * 2) 25 and are as follows: 24-34, 35-45, 46-56 ...
+// If height comes in at 35 we use the 24-34 window, up to height 45.
+// If height comes in at 46 we use the 35-45 window, up to height 56 etc.
+//
+// This function MUST be called with the chain state lock held (for writes).
+func (b *BlockChain) isVoterMajorityVersion(minVer uint32, prevNode *blockNode) bool {
+	// Walk blockchain backwards to calculate version.
+	node, err := b.findStakeVersionPriorNode(prevNode)
+	if err != nil {
+		if err == errStakeVersionHeight {
+			return 0 >= minVer
+		}
+		return false
+	}
+
+	// Calculate version.
```

Comment:
I don't think this comment really provides any useful information, but more importantly, it is also inaccurate.  There is no version calculation going on.  Instead it is tallying both the total number of votes and those with versions >= to the requested minimum version.

I would suggest something like the following:

```Go
// Tally both the total number of votes in the previous stake version validation
// interval and how many of those votes are at least the requested minimum
// version.
```